### PR TITLE
Add validation and default ensembling job values.

### PR DESCRIPTION
**Pull request by ashwinath**
This PR adds the following:

1. Default Spark config annotations that is configurable in the Turing API config file.
2. Add in validation for the Create ensembling job POST body.
3. Nasty side effects include the rewiring of viper configuration delimiter from `.` to `::` due to a problem in unmarshalling strings.

For example:
```go
type Foo struct {
  Bar map[string]string
}
```

with a configuration of

```yaml
Foo:
  Bar:
    "hello.one.two": "some string"
```

Will yield this structure instead:
```foo
Foo = interface{}{
  "Bar": interface{}{
    "hello": interface{}{
      "one": inteface{}{
        "two": "some string",
      },
    },
  },
}
```

To solve this, we change the delimiter to something uncommon, say a cons operator `::`.

### CODE REVIEW

**Comment by krithika369** [comment]
"Nasty side effects ..."

I don't think it's nasty at all - it sounds like a common use case and a legit solution. 😄

File: api/turing/config/config.go [hunk 1]
```diff
@@ -134,6 +135,17 @@ type EnsemblingJobConfig struct {
 	KanikoConfig KanikoConfig `validate:"required"`
 	// ImageBuilderConfig contains the configuration related to the built ensembler image itself.
 	ImageBuilderConfig ImageBuilderConfig `validate:"required"`
+	// DefaultConfigurations contains the default configurations applied to the ensembling job.
+	// The user is free to override/append to the configuration
```
Comment (krithika369) [inline api/turing/config/config.go:139]: > The user is free to override/append to the configuration

By append/override, do we mean the user-supplied config values on app startup or some other way such as configs specified during job submission? I imagine it's the latter and if so, can it be clarified with a line or two in the comments?

File: api/turing/models/ensembling_job.go [hunk 2]
```diff
@@ -76,11 +76,11 @@ func (r *InfraConfig) Scan(value interface{}) error {
 
 // BatchEnsemblingJobResources contains the resources delared to run the ensembling job.
 type BatchEnsemblingJobResources struct {
-	DriverCPURequest      string `json:"driver_cpu_request,omitempty"`
```
Comment (krithika369) [inline api/turing/models/ensembling_job.go:79]: Why has the `omitempty` been removed?

File: api/turing/models/ensembling_job.go [hunk 3]
```diff
@@ -56,8 +56,8 @@ func (r *JobConfig) Scan(value interface{}) error {
 type InfraConfig struct {
 	ArtifactURI        string                       `json:"artifact_uri"`
 	EnsemblerName      string                       `json:"ensembler_name"`
-	ServiceAccountName string                       `json:"service_account_name"`
-	Resources          *BatchEnsemblingJobResources `json:"resources"`
+	ServiceAccountName string                       `json:"service_account_name" validate:"required"`
+	Resources          *BatchEnsemblingJobResources `json:"resources" validate:"required"`
```
Comment (krithika369) [inline api/turing/models/ensembling_job.go:60]: Why is this required but the fields within the `BatchEnsemblingJobResources` are all optional?

File: api/turing/models/ensembling_job.go [hunk 4]
```diff
@@ -12,12 +12,12 @@ import (
 // EnsemblingJob holds the information required for an ensembling job to be done asynchronously
 type EnsemblingJob struct {
 	Model
-	Name            string       `json:"name"`
-	EnsemblerID     ID           `json:"ensembler_id"`
+	Name            string       `json:"name" validate:"required"`
+	EnsemblerID     ID           `json:"ensembler_id" validate:"required"`
 	ProjectID       ID           `json:"project_id"`
 	EnvironmentName string       `json:"environment_name"`
```
Comment (krithika369) [inline api/turing/models/ensembling_job.go:18]: If the environment is simply going to be taken from the Ensembler, does it need to be stored here?

File: api/turing/models/ensembling_job.go (hunk 2 above)
Comment (ashwinath) [reply api/turing/models/ensembling_job.go:79]: I don't think there is a reason to do that anymore since the values are going to be populated

File: api/turing/models/ensembling_job.go (hunk 3 above)
Comment (ashwinath) [reply api/turing/models/ensembling_job.go:60]: This is a mistake, will remove this

File: api/turing/models/ensembling_job.go (hunk 4 above)
Comment (ashwinath) [reply api/turing/models/ensembling_job.go:18]: Not quite related to this PR but I do agree with you. I think it was there last time because I thought it was a user supplied value, but during the review process this was changed to a configuration value and this was left behind.

Comment (ashwinath) [reply api/turing/models/ensembling_job.go:18]: Ah seems like I missed out an important bit of information. There might be a chance next time where this option might be fillable by the user. It would be easier to state the default environment for now in the database as compared to doing a data patch in the future (which might be complicated because the default environment is pulled from MLP).

**Review by krithika369 — APPROVED** [review]
LGTM!
